R/pairwise-comparisons.R

```diff
@@ -0,0 +1,244 @@
+#' @title Do pairwise Comparisons of Scores
+#'
+#' @description
+#'
+#' Make pairwise comparisons between models. The code for the pairwise
+#' comparisons is inspired by an implementation by Johannes Bracher.
+#'
+#' The implementation of the permutation test follows the function
+#' `permuationTest` from the `surveillance` package by Michael Höhle,
+#' Andrea Riebler and Michaela Paul.
+#'
+#' @param scores A data.frame of unsummarised scores as produced by
+#' \code{\link{eval_forecasts}}
+#' @param var string that denotes the column name with the models or forecasters
+#' to compare. Will probably be changed in the future such that a column
+#' named "model" must be present.
+#' @param metric A string with the metric to do the comparison on.
+#' @param permutation_test whether or not to do a permutation test alongside
+#' the comparison. Default is FALSE
+#' @param baseline the baseline model against which to compare other models.
+#' @return A ggplot2 object with a coloured table of summarised scores
+#' @importFrom data.table as.data.table data.table setnames copy
+#' @importFrom stats sd rbinom
+#' @importFrom utils combn
+#' @export
+#' @author Johannes Bracher, https://jbracher.github.io/
+#' @author Nikos Bosse
+#'
+#' @examples
+#'
+#'
+#' df <- data.frame(model = rep(c("model1", "model2", "model3"), each = 10),
+#'                  id = rep(1:10),
+#'                  interval_score = (abs(rnorm(30))),
+#'                  aem = (abs(rnorm(30))))
+#'
+#' res <- pairwise_comparison(df, permutation_test = TRUE, baseline = "model1")
+#' plot_pairwise_comparison(res, type = "ratio")
+#'
+
+
+pairwise_comparison <- function(scores,
+                               var = "model", # alternatively (and probably better): just force user to have a column called "model" or "forecaster"
+                               metric = "interval_score", # maybe the default can happen automatically
+                               permutation_test = FALSE,
+                               # also need to add arguments for the permutation test, i.e. ratio and number of permutations to do
+                               baseline = NULL) { # comparison against a baseline model yet to be implemented
+
+  # get list of models
+  identifiers <- unique(scores[[var]])
+
+  # identify unit of single observation
+  # this basically recovers what was used in eval_forecasts as by argument
+  # could think about adding a specific by argument to this function, but
+  # really this should not be necessary I think
+  all_metrics <- list_of_avail_metrics()
+  by <- setdiff(names(scores), c(all_metrics, var))
+
+  # for efficiency, remove everything that is not in c(by, var)
+  scores <- data.table::as.data.table(scores)
+  cols_to_remove <- setdiff(names(scores), c(by, var, metric))
+  scores[, eval(cols_to_remove) := NULL]
+
+
+  single_comparison <- function(model1, model2, permutation_test = FALSE) {
```

Comment:
ideally have nested functions as internal functions stored elsewhere below in this script rather than inside another function. Minor.

Reply:
this allows you to unit test it etc.